File: app/src/main/java/kfzkennzeichen/gisbertamm/de/kfz_kennzeichen/ResultFragment.java
package kfzkennzeichen.gisbertamm.de.kfz_kennzeichen;

import android.app.AlertDialog;
import android.app.Fragment;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.Base64;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import org.apache.http.HttpResponse;
import org.apache.http.NameValuePair;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.message.BasicNameValuePair;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.List;

import kfzkennzeichen.gisbertamm.de.kfz_kennzeichen.persistence.DatabaseHandler;
import kfzkennzeichen.gisbertamm.de.kfz_kennzeichen.persistence.SavedEntry;

public class ResultFragment extends Fragment {

    public static final String ENTRY = "entry";

    public static ResultFragment newInstance(SavedEntry entry) {
        ResultFragment f = new ResultFragment();

        Bundle args = new Bundle();
        args.putSerializable(ENTRY, entry);
        f.setArguments(args);

        return f;
    }

    public SavedEntry getSavedEntry() {
        return (SavedEntry) getArguments().getSerializable(ENTRY);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_result, container, false);
        final SavedEntry savedEntry = getSavedEntry();
        final TextView code = (TextView) view.findViewById(R.id.result_code);
        Button proposeJoke = (Button) view.findViewById(R.id.button_propose_joke);

        if (savedEntry == null) { // nothing found
            code.setText(getString(R.string.nothing_found));
            proposeJoke.setVisibility(View.INVISIBLE);
            view.findViewById(R.id.result_jokes_heading).setVisibility(View.INVISIBLE);
            return view;
        }

        TextView district = (TextView) view.findViewById(R.id.result_district);
        TextView districtCenter = (TextView) view.findViewById(R.id.result_district_center);
        TextView jokes = (TextView) view.findViewById(R.id.result_jokes);
        ImageView crestView = (ImageView) view.findViewById(R.id.crest);

        code.setText(savedEntry.getCode());

        district.setText(savedEntry.getDistrict());
        district.setClickable(true);
        district.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                openWikipediaPage(savedEntry);
            }
        });
        districtCenter.setText("abgeleitet von: " + savedEntry.getDistrictCenter()
                + " (" + savedEntry.get_state() + ")");

        String jokesText = concatJokesStrings(savedEntry.getJokes());
        jokes.setText(jokesText);

        proposeJoke.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Set an EditText view to get user input
                final EditText input = new EditText(getActivity());

                new AlertDialog.Builder(getActivity())
                        .setTitle("Eigenen Text vorschlagen")
                        .setMessage("Bitte Text eingeben (maximal 30 Zeichen).")
                        .setView(input)
                        .setPositiveButton("Vorschlagen", new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int whichButton) {
                                MailProposalTask proposalTask = new MailProposalTask();
                                proposalTask.execute(code.getText().toString(), input.getText().toString());
                            }
                        })
                        .setNegativeButton("Abbrechen", new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int whichButton) {
                                // Do nothing.
                            }
                        }).show();
            }
        });

        final int resourceId = getResources().getIdentifier(savedEntry.getCrestIdentifier(), "drawable",
                getActivity().getPackageName());
        try {
            Drawable d = getResources().getDrawable(resourceId);
            if (d != null) {
                crestView.setImageDrawable(d);
            } else {
                Log.e(getClass().getSimpleName(), "Crest drawable is null for " + savedEntry.getCode() + ", identifier: " + savedEntry.getCrestIdentifier());
            }
        } catch (Resources.NotFoundException e) {
            Log.e(getClass().getSimpleName(), "Crest resource not found for " + savedEntry.getCode() + ", identifier: " + savedEntry.getCrestIdentifier());
        }

        crestView.setClickable(true);
        crestView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                openWikipediaPage(savedEntry);
            }
        });

        return view;
    }

    private void openWikipediaPage(SavedEntry savedEntry) {
        String url = DatabaseHandler.WIKIPEDIA_BASE_URL + savedEntry.getDistrictWikipediaUrl();
        Log.d(getClass().getSimpleName(), "Open url: " + url);
        Intent i = new Intent(Intent.ACTION_VIEW);
        i.setData(Uri.parse(url));
        startActivity(i);
    }

    private String concatJokesStrings(List<String> jokesList) {
        StringBuilder jokesBuff = new StringBuilder();
        for (String joke : jokesList) {
            jokesBuff.append(joke);
            jokesBuff.append("\n");
        }
        return jokesBuff.toString();
    }

    private class MailProposalTask extends AsyncTask<String, Void, Result> {

        @Override
        protected Result doInBackground(String... params) {
            HttpClient httpClient = new DefaultHttpClient();

            HttpPost httpPost = new HttpPost("https://api.mailgun.net/v3/sandbox47fa9b0a752440c794641c362d468402.mailgun.org/messages");

            String base64EncodedCredentials = "Basic " + Base64.encodeToString(
                    ("api" + ":" + getResources().getString(R.string.api_key)).getBytes(),
                    Base64.NO_WRAP);


            httpPost.setHeader("Authorization", base64EncodedCredentials);

            List<NameValuePair> nameValuePairList = new ArrayList<NameValuePair>();
            BasicNameValuePair fromBasicNameValuePair = new BasicNameValuePair("from", "In-App-Proposal <kfz-kennzeichen-spruch-vorschlag@web.de>");
            nameValuePairList.add(fromBasicNameValuePair);
            BasicNameValuePair toBasicNameValuePAir = new BasicNameValuePair("to", "kfz-kennzeichen-spruch-vorschlag@web.de");
            nameValuePairList.add(toBasicNameValuePAir);
            BasicNameValuePair subjectBasicNameValuePair = new BasicNameValuePair("subject", "Neuer Vorschlag für Kennzeichen-Joke (Android)");
            nameValuePairList.add(subjectBasicNameValuePair);
            BasicNameValuePair textBasicNameValuePAir = new BasicNameValuePair("text", "code: " + params[0] + ", Vorschlag: " + params[1]);
            nameValuePairList.add(textBasicNameValuePAir);

            try {
                UrlEncodedFormEntity urlEncodedFormEntity = new UrlEncodedFormEntity(nameValuePairList, "UTF-8");

                httpPost.setEntity(urlEncodedFormEntity);

                try {
                    HttpResponse httpResponse = httpClient.execute(httpPost);

                    Result result = new Result();
                    result.statuscode = httpResponse.getStatusLine().getStatusCode();
                    result.statusLine = httpResponse.getStatusLine().getReasonPhrase();

                    if (httpResponse.getStatusLine().getStatusCode() != 200) {
                        return result;
                    }

                    InputStream inputStream = httpResponse.getEntity().getContent();

                    InputStreamReader inputStreamReader = new InputStreamReader(inputStream);

                    BufferedReader bufferedReader = new BufferedReader(inputStreamReader);

                    StringBuilder stringBuilder = new StringBuilder();

                    String bufferedStrChunk = null;

                    while ((bufferedStrChunk = bufferedReader.readLine()) != null) {
                        stringBuilder.append(bufferedStrChunk);
                    }

                    result.body = stringBuilder.toString();
                    return result;

                } catch (ClientProtocolException cpe) {
                    Log.e(getClass().getSimpleName(), cpe.getLocalizedMessage(), cpe);
                    Result result = new Result();
                    result.statuscode = -1;
                    result.body = cpe.getLocalizedMessage();
                    return result;
                } catch (IOException ioe) {
                    Log.e(getClass().getSimpleName(), ioe.getLocalizedMessage(), ioe);
                    Result result = new Result();
                    result.statuscode = -1;
                    result.body = ioe.getLocalizedMessage();
                    return result;
                }

            } catch (UnsupportedEncodingException uee) {
                Log.e(getClass().getSimpleName(), "Exception given because of UrlEncodedFormEntity argument: code: "
                        + params[0] + ", Vorschlag: " + params[1], uee);
                Result result = new Result();
                result.statuscode = -1;
                result.body = uee.getLocalizedMessage();
                return result;
            }
        }

        @Override
        protected void onPostExecute(Result result) {
            super.onPostExecute(result);

            if (result.statuscode == 200) {
                Toast.makeText(getActivity().getApplicationContext(), "Der Vorschlag wurde erfolgreich übermittelt.", Toast.LENGTH_SHORT).show();
                Log.d(getClass().getSimpleName(), result.body);
            } else {
                // HTTP error
                String errorMessage = result.statuscode + "(" + result.statusLine + ")";
                if (result.statuscode == -1) { // Exception
                    errorMessage = result.body;
                }
                Toast.makeText(getActivity().getApplicationContext(), "Der Vorschlag konnte leider nicht übermittelt werden. Fehler: "
                        + errorMessage, Toast.LENGTH_SHORT).show();
            }
        }
    }

    private class Result {
        int statuscode;
        String statusLine;
        String body;
    }
}

File: app/src/main/java/kfzkennzeichen/gisbertamm/de/kfz_kennzeichen/ResultFragment.kt
package kfzkennzeichen.gisbertamm.de.kfz_kennzeichen

import android.app.AlertDialog
import android.app.Fragment
import android.content.Intent
import android.content.res.Resources.NotFoundException
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.text.InputFilter
import android.text.InputFilter.LengthFilter
import android.util.Base64
import android.util.Log
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import kfzkennzeichen.gisbertamm.de.kfz_kennzeichen.persistence.DatabaseHandler
import kfzkennzeichen.gisbertamm.de.kfz_kennzeichen.persistence.SavedEntry
import org.apache.http.NameValuePair
import org.apache.http.client.ClientProtocolException
import org.apache.http.client.HttpClient
import org.apache.http.client.entity.UrlEncodedFormEntity
import org.apache.http.client.methods.HttpPost
import org.apache.http.impl.client.DefaultHttpClient
import org.apache.http.message.BasicNameValuePair
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.UnsupportedEncodingException
import java.util.*

class ResultFragment : Fragment() {
    val savedEntry: SavedEntry?
        get() = arguments.getSerializable(ENTRY) as SavedEntry?

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle): View? {
        val view = inflater.inflate(R.layout.fragment_result, container, false)
        val savedEntry = savedEntry
        val code = view.findViewById<View>(R.id.result_code) as TextView
        val proposeJoke = view.findViewById<View>(R.id.button_propose_joke) as Button
        if (savedEntry == null) { // nothing found
            code.text = getString(R.string.nothing_found)
            proposeJoke.visibility = View.INVISIBLE
            view.findViewById<View>(R.id.result_jokes_heading).visibility = View.INVISIBLE
            return view
        }
        val district = view.findViewById<View>(R.id.result_district) as TextView
        val districtCenter = view.findViewById<View>(R.id.result_district_center) as TextView
        val jokes = view.findViewById<View>(R.id.result_jokes) as TextView
        val crestView = view.findViewById<View>(R.id.crest) as ImageView
        code.text = savedEntry.code
        district.text = savedEntry.district
        district.isClickable = true
        district.setOnClickListener { openWikipediaPage(savedEntry) }
        districtCenter.text = ("abgeleitet von: " + savedEntry.districtCenter
                + " (" + savedEntry._state + ")")
        val jokesText = concatJokesStrings(savedEntry.jokes)
        jokes.text = jokesText
        proposeJoke.setOnClickListener {
            // set an EditText view to get user input; need layout around for padding
            val layout = LinearLayout(activity)
            layout.orientation = LinearLayout.VERTICAL
            layout.gravity = Gravity.CENTER_HORIZONTAL
            val input = EditText(activity)
            val maxLength = 30
            input.filters = arrayOf<InputFilter>(LengthFilter(maxLength))
            layout.setPadding(20, 0, 20, 0)
            layout.addView(input)
            AlertDialog.Builder(activity)
                    .setTitle("Eigenen Text vorschlagen")
                    .setMessage("Bitte Text eingeben (maximal 30 Zeichen).")
                    .setView(layout)
                    .setPositiveButton("Vorschlagen") { dialog, whichButton ->
                        val proposalTask = MailProposalTask()
                        proposalTask.execute(code.text.toString(), input.text.toString())
                    }
                    .setNegativeButton("Abbrechen") { dialog, whichButton ->
                        // Do nothing.
                    }.show()
        }
        val resourceId = resources.getIdentifier(savedEntry.crestIdentifier, "drawable",
                activity.packageName)
        try {
            val d = resources.getDrawable(resourceId)
            if (d != null) {
                crestView.setImageDrawable(d)
            } else {
                Log.e(javaClass.simpleName, "Crest drawable is null for " + savedEntry.code + ", identifier: " + savedEntry.crestIdentifier)
            }
        } catch (e: NotFoundException) {
            Log.e(javaClass.simpleName, "Crest resource not found for " + savedEntry.code + ", identifier: " + savedEntry.crestIdentifier)
        }
        crestView.isClickable = true
        crestView.setOnClickListener { openWikipediaPage(savedEntry) }
        return view
    }

    private fun openWikipediaPage(savedEntry: SavedEntry) {
        val url = DatabaseHandler.WIKIPEDIA_BASE_URL + savedEntry.districtWikipediaUrl
        Log.d(javaClass.simpleName, "Open url: $url")
        val i = Intent(Intent.ACTION_VIEW)
        i.data = Uri.parse(url)
        startActivity(i)
    }

    private fun concatJokesStrings(jokesList: List<String>): String {
        val jokesBuff = StringBuilder()
        for (joke in jokesList) {
            jokesBuff.append(joke)
            jokesBuff.append("\n")
        }
        return jokesBuff.toString()
    }

    private inner class MailProposalTask : AsyncTask<String?, Void?, Result>() {
        protected override fun doInBackground(vararg params: String): Result {
            val httpClient: HttpClient = DefaultHttpClient()
            val httpPost = HttpPost("https://api.mailgun.net/v3/sandbox47fa9b0a752440c794641c362d468402.mailgun.org/messages")
            val base64EncodedCredentials = "Basic " + Base64.encodeToString(
                    ("api" + ":" + resources.getString(R.string.api_key)).toByteArray(),
                    Base64.NO_WRAP)
            httpPost.setHeader("Authorization", base64EncodedCredentials)
            val nameValuePairList: MutableList<NameValuePair> = ArrayList()
            val fromBasicNameValuePair = BasicNameValuePair("from", "In-App-Proposal <kfz-kennzeichen-spruch-vorschlag@web.de>")
            nameValuePairList.add(fromBasicNameValuePair)
            val toBasicNameValuePAir = BasicNameValuePair("to", "kfz-kennzeichen-spruch-vorschlag@web.de")
            nameValuePairList.add(toBasicNameValuePAir)
            val subjectBasicNameValuePair = BasicNameValuePair("subject", "Neuer Vorschlag für Kennzeichen-Joke (Android)")
            nameValuePairList.add(subjectBasicNameValuePair)
            val textBasicNameValuePAir = BasicNameValuePair("text", params[0] + "; " + params[1])
            nameValuePairList.add(textBasicNameValuePAir)
            return try {
                val urlEncodedFormEntity = UrlEncodedFormEntity(nameValuePairList, "UTF-8")
                httpPost.entity = urlEncodedFormEntity
                try {
                    val httpResponse = httpClient.execute(httpPost)
                    val result: Result = Result()
                    result.statuscode = httpResponse.statusLine.statusCode
                    result.statusLine = httpResponse.statusLine.reasonPhrase
                    if (httpResponse.statusLine.statusCode != 200) {
                        return result
                    }
                    val inputStream = httpResponse.entity.content
                    val inputStreamReader = InputStreamReader(inputStream)
                    val bufferedReader = BufferedReader(inputStreamReader)
                    val stringBuilder = StringBuilder()
                    var bufferedStrChunk: String? = null
                    while (bufferedReader.readLine().also { bufferedStrChunk = it } != null) {
                        stringBuilder.append(bufferedStrChunk)
                    }
                    result.body = stringBuilder.toString()
                    result
                } catch (cpe: ClientProtocolException) {
                    Log.e(javaClass.simpleName, cpe.localizedMessage, cpe)
                    val result: Result = Result()
                    result.statuscode = -1
                    result.body = cpe.localizedMessage
                    result
                } catch (ioe: IOException) {
                    Log.e(javaClass.simpleName, ioe.localizedMessage, ioe)
                    val result: Result = Result()
                    result.statuscode = -1
                    result.body = ioe.localizedMessage
                    result
                }
            } catch (uee: UnsupportedEncodingException) {
                Log.e(javaClass.simpleName, "Exception given because of UrlEncodedFormEntity argument: code: "
                        + params[0] + ", Vorschlag: " + params[1], uee)
                val result: Result = Result()
                result.statuscode = -1
                result.body = uee.localizedMessage
                result
            }
        }

        override fun onPostExecute(result: Result) {
            super.onPostExecute(result)
            if (result.statuscode == 200) {
                Toast.makeText(activity.applicationContext, "Der Vorschlag wurde erfolgreich übermittelt.", Toast.LENGTH_SHORT).show()
                Log.d(javaClass.simpleName, result.body)
            } else {
                // HTTP error
                var errorMessage: String? = result.statuscode.toString() + "(" + result.statusLine + ")"
                if (result.statuscode == -1) { // Exception
                    errorMessage = result.body
                }
                Toast.makeText(activity.applicationContext, "Der Vorschlag konnte leider nicht übermittelt werden. Fehler: "
                        + errorMessage, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private inner class Result {
        var statuscode = 0
        var statusLine: String? = null
        var body: String? = null
    }

    companion object {
        const val ENTRY = "entry"
        fun newInstance(entry: SavedEntry?): ResultFragment {
            val f = ResultFragment()
            val args = Bundle()
            args.putSerializable(ENTRY, entry)
            f.arguments = args
            return f
        }
    }
}